tests/csapi/sync_test.go

```diff
@@ -731,6 +731,163 @@ func TestRoomSummary(t *testing.T) {
+	bobSince := bob.MustSyncUntil(t, client.SyncReq{}, client.SyncJoinedTo(bob.UserID, roomID))
+
+	// 3. Bob leaves the room
+	bob.MustLeaveRoom(t, roomID)
+
+	// 4. Bob rejoins the room
+	bob.MustJoinRoom(t, roomID, nil)
+
+	// 5. Bob leaves the room again
+	bob.MustLeaveRoom(t, roomID)
+
+	// 6. Bob does a sync
+	jsonRes, _ := bob.MustSync(t, client.SyncReq{Since: bobSince})
+
+	// Bob should only see a leave event, not both a join and a leave event
+	if !jsonRes.Get("rooms.leave." + client.GjsonEscape(roomID)).Exists() {
+		t.Errorf("Expected to see the room in the leave section of the sync response")
+	}
+
+	// Make sure there's no join event for the room
+	if jsonRes.Get("rooms.join." + client.GjsonEscape(roomID)).Exists() {
+		t.Errorf("Room should not appear in the join section of the sync response")
+	}
+}
+
+// TestLeaveReinviteSyncFederated tests that when a user is kicked and then re-invited over federation,
+// they only see an invite in their sync response, not both an invite and leave event.
+func TestLeaveReinviteSyncFederated(t *testing.T) {
+	deployment := complement.Deploy(t, 2)
+	defer deployment.Destroy(t)
+
+	alice := deployment.Register(t, "hs1", helpers.RegistrationOpts{})
+	bob := deployment.Register(t, "hs2", helpers.RegistrationOpts{})
+	
+	// Charlie is simply here to flag when events arrive over federation to hs2.
+	charlie := deployment.Register(t, "hs2", helpers.RegistrationOpts{})
+
+	// 1. Alice creates a room and both Bob and Charlie join
+	roomID := alice.MustCreateRoom(t, map[string]any{
+		"preset": "public_chat",
+	})
+
+	// Users on hs2 need to join via federation, so we need to specify the server name
+	alice.MustInviteRoom(t, roomID, bob.UserID)
+	alice.MustInviteRoom(t, roomID, charlie.UserID)
+	bob.MustSyncUntil(t, client.SyncReq{}, client.SyncInvitedTo(bob.UserID, roomID))
+	charlie.MustSyncUntil(t, client.SyncReq{}, client.SyncInvitedTo(charlie.UserID, roomID))
+
+	bob.MustJoinRoom(t, roomID, []spec.ServerName{
+		deployment.GetFullyQualifiedHomeserverName(t, "hs1"),
+	})
+	charlie.MustJoinRoom(t, roomID, []spec.ServerName{
+		deployment.GetFullyQualifiedHomeserverName(t, "hs1"),
+	})
+
+	// 2. Both Bob and Charlie do a sync and verify that they see the join
+	bobSince := bob.MustSyncUntil(t, client.SyncReq{}, client.SyncJoinedTo(bob.UserID, roomID))
+	charlieSince := charlie.MustSyncUntil(t, client.SyncReq{}, client.SyncJoinedTo(charlie.UserID, roomID))
+
+	// 3. Alice kicks Bob from the room and then re-invites them.
+	// For kicking, we need to use the POST /rooms/{roomId}/kick endpoint
+	alice.MustDo(t, "POST", []string{"_matrix", "client", "v3", "rooms", roomID, "kick"},
+		client.WithJSONBody(t, map[string]any{
+			"user_id": bob.UserID,
+		}),
+	)
+
+	// Wait until Bob is kicked.
+	aliceSince := alice.MustSyncUntil(t, client.SyncReq{}, client.SyncLeftFrom(bob.UserID, roomID))
+	charlieSince = charlie.MustSyncUntil(t, client.SyncReq{Since: charlieSince}, client.SyncLeftFrom(bob.UserID, roomID))
```

Comment:
I don't think we need to worry about getting a `since` token in any of these places execpt for `bob` which we are actually testing incremental sync for.

The other sync's are just asserting pre-conditions.

```suggestion
	charlie.MustSyncUntil(t, client.SyncReq{}, client.SyncLeftFrom(bob.UserID, roomID))
```